File: 702/memserver/dump_module.py
import sys
import threading
import signal
import os
from client import read_mem, read_ptr, tarea

ourlogo = """
         ___      _                _           _                                 
   __ _ / _ \ ___| |__   __ _ _ __( )___    __| |_   _ _ __ ___  _ __   ___ _ __ 
  / _` | | | |_  / '_ \ / _` | '__|// __|  / _` | | | | '_ ` _ \| '_ \ / _ \ '__|
 | (_| | |_| |/ /| | | | (_| | |    \__ \ | (_| | |_| | | | | | | |_) |  __/ |   
  \__,_|\___//___|_| |_|\__,_|_|    |___/  \__,_|\__,_|_| |_| |_| .__/ \___|_|   
    For dumping modules needed to build ropchains               |_|              
"""

print(ourlogo)
print('Instructions:')
print('Navigate the PS4 web browser to port 8080 on this PC')
print('(or just press OK if you are already on error screen)')
print('When you hit the error screen (again), press Enter on this PC.')


some_func = read_ptr(read_ptr(read_ptr(tarea+0x18)))

if ',' in sys.argv[1]:
    idx = int(sys.argv[1].split(',')[0])
    offset0 = int(sys.argv[1].split(',')[1], 16)
else:
    idx = int(sys.argv[1])
    offset0 = 0

if idx > 0:
    plt = some_func - 10117000
    plt_entry = plt + idx * 16
    q = read_mem(plt_entry, 6)
    assert q[:2] == b'\xff%', q
    got_entry = plt_entry + 6 + int.from_bytes(q[2:], 'little')
    got_func = read_ptr(got_entry)
else:
    got_func = some_func

got_func += offset0
data = b''
chunk_sz = int(os.environ.get('CHUNK_SIZE', '4096'))

def watchdog_thread():
    input()
    os.kill(os.getpid(), signal.SIGINT)

threading.Thread(target=watchdog_thread, daemon=True).start()

try:
    while True:
        data += read_mem(got_func + len(data), chunk_sz)
        # if chunk_sz < 1048576:
        #    chunk_sz *= 2
        print(len(data), end=' bytes loaded\r')
except KeyboardInterrupt:
    print('\nKeyboardInterrupt')

if len(sys.argv) > 2:
    print('saving to', sys.argv[2])
    with open(sys.argv[2], 'wb') as file: 
        file.write(data)

File: JSVersions/memserver/dump_module.js
const fs = require('fs');
const readline = require('readline');

console.log('Navigate the PS4 web browser to port 8080 on this PC');
console.log('(or just press OK if you are already on error screen)');
console.log('When you hit the error screen (again), press Enter on this PC.');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

const client = require('./client.js');
const { read_mem, read_ptr, tarea } = client;

const some_func = read_ptr(read_ptr(read_ptr(tarea + 0x18)));

let idx, offset0;

if (process.argv[2].includes(',')) {
    idx = parseInt(process.argv[2].split(',')[0]);
    offset0 = parseInt(process.argv[2].split(',')[1], 16);
} else {
    idx = parseInt(process.argv[2]);
    offset0 = 0;
}

let got_func;

if (idx < 0) {
    got_func = some_func;
} else {
    const plt = some_func - 10063176;
    const plt_entry = plt + idx * 16;
    const q = read_mem(plt_entry, 6);
    if (!q.slice(0, 2).equals(Buffer.from([0xff, 0x25]))) {
        throw new Error("Assertion failed");
    }
    const got_entry = plt_entry + 6 + q.readUIntLE(2, 4);
    got_func = read_ptr(got_entry);
}

got_func += offset0;

let data = Buffer.alloc(0);
let chunk_sz = 1;

const watchdog_thread = () => {
    rl.question('', () => {
        process.kill(process.pid, 'SIGINT');
    });
};

setInterval(watchdog_thread, 1000);

try {
    while (true) {
        data = Buffer.concat([data, read_mem(got_func + data.length, chunk_sz)]);
        if (chunk_sz < 1024) {
            chunk_sz *= 2;
        }
        process.stdout.write(`${data.length} bytes loaded\r`);
    }
} catch (error) {
    if (error.message !== 'SIGINT') {
        console.log('\n' + error.message);
    }
}

if (process.argv.length > 3) {
    console.log('saving to', process.argv[3]);
    fs.writeFileSync(process.argv[3], data);
}
